module: add module.detectSyntax (#57731)
nodejs/node: doc/api/module.md

const siblingModule = require('./sibling-module');
```

### `module.detectSyntax(code)`

<!-- YAML
added: REPLACEME
-->

> Stability: 1.0 - Early development

> The `module.detectSyntax()` method attempts to determine the syntax of agiven source code string.
> This detection is performed without executing the code, relying on file extensions, or checking the type field in `package.json`.

Reviewer: Typo, and the second sentence is ambiguous.

```suggestion
The `module.detectSyntax()` method attempts to determine the syntax of a given source code string.
This detection is performed only by parsing the source code; there is no evaluation or reading of file extensions
or reading of `package.json` files.
```

Only TypeScript syntax that can be erased is supported.

* `code` {string} The code to detect the syntax of.
* Returns: {string|undefined} The detected syntax of the code. Possible values are:
  * `'commonjs'` for CommonJS module code.
  * `'module'` for ECMAScript module code.
  * `'commonjs-typescript'` for CommonJS TypeScript code.
  * `'module-typescript'` for ECMAScript TypeScript code.
  * `undefined` if the syntax cannot be detected.
    **Default:** `'commonjs'`.

```mjs
import { detectSyntax } from 'node:module';
detectSyntax('export const a = 1;'); // 'module'
detectSyntax('const a = 1;'); // 'commonjs'
detectSyntax('export const a: number = 1;'); // 'module-typescript'
detectSyntax('const a: number = 1;'); // 'commonjs-typescript'
detectSyntax('const foo;'); // Invalid syntax, returns undefined
```

There are some limitations to the detection, where the syntax cannot be determined because it's ambiguous:

```mjs
import { detectSyntax } from 'node:module';
// This could be CommonJS if we interpret < and > as greater and lesser,
// or it could be TypeScript if we interpret them as type parameters.

> // In this case, it will be interpreted as TypeScript.

Reviewer 2: this seems potentially problematic, since TS isn't a syntactic superset of JS. Perhaps in this case it could return an array of two choices?

Author: it's an edge case, how do you detect it though 😄 I guess it has to be clear its heuristic, it trades some correctness for convenience

Reviewer 2: if it can be parsed as both, then it's the edge case.

since this function is intended to be used by libraries and not by most users, it seems like correctness is the better tradeoff to make here. Doing `result?.[0] ?? result` isn't that inconvenient, but allows the choice and visibility into the edge cases.

Author: the only way I see to disambiguate the syntax is to `compileScript` and see if it works, and attempt to strip types and if the output is different, it means its an edge case.

Reviewer 2: Sure, that seems fine to me.

```

### `module.detectSyntax(code)`

<!-- YAML
added: REPLACEME
-->

> Stability: 1.0 - Early development

The `module.detectSyntax()` method attempts to determine the syntax of a given source code string.
This detection is performed only by parsing the source code; there is no evaluation or reading of file extensions
or reading of `package.json` files.
Only TypeScript syntax that can be erased is supported.

* `code` {string} The code to detect the syntax of.
* Returns: {string|undefined} The detected syntax of the code. Possible values are:
  * `'commonjs'` for CommonJS module code.
  * `'module'` for ECMAScript module code.
  * `'commonjs-typescript'` for CommonJS TypeScript code.
  * `'module-typescript'` for ECMAScript TypeScript code.
  * `undefined` if the syntax cannot be detected.
    **Default:** `'commonjs'`.

```mjs
import { detectSyntax } from 'node:module';
detectSyntax('export const a = 1;'); // 'module'
detectSyntax('const a = 1;'); // 'commonjs'
detectSyntax('export const a: number = 1;'); // 'module-typescript'
detectSyntax('const a: number = 1;'); // 'commonjs-typescript'
detectSyntax('const foo;'); // Invalid syntax, returns undefined

> ```

Reviewer 3: Maybe this should instead return two booleans where one indicates if the code contained module specific syntax and the other whether it contained typescript syntax? Stating `const a = 1;` is commonjs isn't exactly accurate because that can't be determined solely from the text, as alluded to in the description.
